### PR TITLE
Hub EIA transition

### PR DESCRIPTION
**Pull request by cmgosnell**
**Overall Goal**: The Hub team needs access to the net generation allocated at the generator/fuel level of granularity. They also want this generator/fuel data to be scaled to each of the owners of each of said generators.

**Solution**: For the first part of the above goal... this allocated net gen at the generator/fuel level was an interim output inside of `pudl.analysis.allocate_net_gen.allocate_gen_fuel_by_gen()`. I exposed this interim step through an additional function in `pudl.analysis.allocate_net_gen` AND created a `pudl_out` access to the output.

For the second part of the above goal (the ownership scaling)... we already had code to do this type of thing in `pudl.analysis.plant_parts_eia.MakeMegaGenTbl`. A small tweak was needed to make the existing methods usable for this purpose. I added a very WIP `pudl_out` method called `gen_pm_fuel_ownership` that uses the generator/fuel level output from part one and feeds it into `pudl.analysis.plant_parts_eia.MakeMegaGenTbl.slice_by_ownership()`.

**Question for reviewer**:
- [x] Where should the "work" that is currently happening in `pudl_out.gen_pm_fuel_ownership()` live?? Right now it is very out of step with the rest of `pudl_out` because it is more complex then "run this function"... but I'm not sure where it belongs because it is a strange wrapper or two previously disparate processes.
- [x] Tests? I am mostly wrapping functions/methods that already exist different ways here... but are there any tests (besides just running the  thing in the CI tests) that I should add?

### CODE REVIEW

File: src/pudl/analysis/allocate_net_gen.py [hunk 1]
```diff
@@ -146,25 +176,14 @@ def allocate_gen_fuel_by_gen(pudl_out):
...
-        end_date=pudl_out.end_date,
-    )
-    return gen_allocated
+    gen_pm_fuel = allocate_gen_fuel_by_gen_pm_fuel_guts(gf, gen, gens)
+    _test_gen_fuel_allocation(gen, gen_pm_fuel)
+    return gen_pm_fuel
 
 
-def allocate_gen_fuel_by_gen_pm_fuel(gf, gen, gens, drop_interim_cols=True):
+def allocate_gen_fuel_by_gen_pm_fuel_guts(gf, gen, gens, drop_interim_cols=True):
```
Comment (cmgosnell) [inline src/pudl/analysis/allocate_net_gen.py:186]: I obviously need to rename this.... but I used it's previous name for the wrapper function that grabs the pudl_out tables and then does this stuff. suggestions?

File: src/pudl/analysis/allocate_net_gen.py [hunk 2]
```diff
@@ -105,24 +105,54 @@
 """Data columns from generation_fuel_eia923 that are being allocated."""
 
 
-def allocate_gen_fuel_by_gen(pudl_out):
+def allocate_gen_fuel_by_gen(pudl_out, gen_pm_fuel: pd.DataFrame) -> pd.DataFrame:
```
Comment (cmgosnell) [inline src/pudl/analysis/allocate_net_gen.py:108]: rename to aggregate

File: src/pudl/analysis/allocate_net_gen.py [hunk 3]
```diff
@@ -105,24 +105,54 @@
...
+    gen_allocated = pudl.output.eia923.denorm_generation_eia923(
+        g_df=gen_allocated,
+        pudl_engine=pudl_out.pudl_engine,
+        start_date=pudl_out.start_date,
+        end_date=pudl_out.end_date,
+    )
+    return gen_allocated
+
+
+def allocate_gen_fuel_by_gen_pm_fuel(pudl_out):
```
Comment (cmgosnell) [inline src/pudl/analysis/allocate_net_gen.py:143]: make clear we are allocating net generation and fuel consumption to the generator/pm/fuel type level

**Comment by cmgosnell** [comment]
Will also Close #1549

File: src/pudl/output/pudltabl.py [hunk 4]
```diff
@@ -802,14 +807,40 @@ def gen_original_eia923(self, update=False):
             )
         return self._dfs["gen_og_eia923"]
 
+    def gen_fuel_allocated_generator_fuel_type_eia923(self, update=False):
```
Comment (zaneselvans) [inline src/pudl/output/pudltabl.py:810]: There are so many of these gen/net gen/gen fuel functions now I feel like they need a naming convention, or at least a standard way of describing what the primary key columns are, and where the data columns came from. It still feels like a mix of ad-hoc names. Like I think this one has a PK of `report_date`, `plant_id_eia`, `generator_id`, `prime_mover_code` and `fuel_type_code` (or is it `energy_source_code` now?) but prime mover doesn't show up in the function name, while it is included in `gen_pm_fuel_ownership` below.

Comment (zaneselvans) [inline src/pudl/output/pudltabl.py:810]: Is there fuel consumption also reported in the generation fuel table? What happens to that data in the allocation process? The name makes it sound like maybe that gets allocated too.

File: src/pudl/output/pudltabl.py [hunk 5]
```diff
@@ -802,14 +807,40 @@ def gen_original_eia923(self, update=False):
...
+        """Net generation from gen fuel table allocated to generator/prime_mover/fuel."""
+        if update or self._dfs["gen_pm_fuel_eia923"] is None:
+            self._dfs[
+                "gen_pm_fuel_eia923"
+            ] = allocate_generation_fuel_by_generator_fuel_type(pudl_out=self)
+        return self._dfs["gen_pm_fuel_eia923"]
+
     def gen_allocated_eia923(self, update=False):
```
Comment (zaneselvans) [inline src/pudl/output/pudltabl.py:818]: Now that there's more than one kind of "allocated" generation this function name is ambiguous.

File: src/pudl/analysis/plant_parts_eia.py [hunk 6]
```diff
@@ -415,7 +416,11 @@ def execute(
             mcoe: generator-based mcoe table from :meth:`pudl.output.PudlTabl.mcoe()`
             own_eia860: ownership table from :meth:`pudl.output.PudlTabl.own_eia860()`
             slice_cols: list of columns to slice by ownership fraction in
-                :meth:`MakeMegaGenTbl.slice_by_ownership`. Default is :py:const:`SUM_COLS`
+                :meth:`MakeMegaGenTbl.scale_by_ownership`. Default is :py:const:`SUM_COLS`
+            validate_own_merge: how the merge between ``mcoe`` and ``owm_eia860``
```
Comment (zaneselvans) [inline src/pudl/analysis/plant_parts_eia.py:420]: `owm_eia860` should be `own_eia860`

File: src/pudl/analysis/plant_parts_eia.py [hunk 7]
```diff
@@ -415,7 +416,11 @@ def execute(
             mcoe: generator-based mcoe table from :meth:`pudl.output.PudlTabl.mcoe()`
             own_eia860: ownership table from :meth:`pudl.output.PudlTabl.own_eia860()`
             slice_cols: list of columns to slice by ownership fraction in
```
Comment (zaneselvans) [inline src/pudl/analysis/plant_parts_eia.py:418]: Should `slice_cols` be `scale_cols`?

File: src/pudl/analysis/allocate_net_gen.py [hunk 8]
```diff
@@ -235,25 +195,101 @@ def allocate_gen_fuel_by_gen_pm_fuel(gf, gen, gens, drop_interim_cols=True):
...
+    pudl_out, gen_pm_fuel: pd.DataFrame
+) -> pd.DataFrame:
+    """
+    Aggregate gen fuel data columns to generators.
+
+    The generation_fuel_eia923 table includes net generation and fuel
+    consumption data at the plant/fuel type/prime mover level. The most
+    granular level of plants that PUDL typically uses is at the plant/generator
+    level. This function takes the plant/fuel type/prime mover level allocation,
+    aggregates it to the generator level and then denormalizes it to make it
```
Comment (zaneselvans) [inline src/pudl/analysis/allocate_net_gen.py:212]: What do you mean by "denormalizes" here?

File: src/pudl/analysis/allocate_net_gen.py [hunk 9]
```diff
@@ -235,25 +195,101 @@ def allocate_gen_fuel_by_gen_pm_fuel(gf, gen, gens, drop_interim_cols=True):
...
+        end_date=pudl_out.end_date,
+    )
+    return gen_allocated
+
+
+def scale_allocated_net_gen_by_ownership(
+    gen_pm_fuel: pd.DataFrame, gens: pd.DataFrame, own_eia860: pd.DataFrame
+) -> pd.DataFrame:
+    """
+    Scale the allocated net generation at the generator/fuel-type level by ownership.
```
Comment (zaneselvans) [inline src/pudl/analysis/allocate_net_gen.py:242]: Is it actually (generator, fuel type) or is it (generator, prime mover, fuel type)?

File: src/pudl/analysis/allocate_net_gen.py [hunk 10]
```diff
@@ -235,25 +195,101 @@ def allocate_gen_fuel_by_gen_pm_fuel(gf, gen, gens, drop_interim_cols=True):
...
+
+    It can be helpful to have a table of net generation and fuel consumption
+    at the generator/fuel-type level (i.e. the result of :func:`allocate_gen_fuel_by_gen_pm_fuel`)
+    to be associated and scaled with all of the owners of those generators.
+    This can be helpful because it can enable a utility aggregation of fuel use.
```
Comment (zaneselvans) [inline src/pudl/analysis/allocate_net_gen.py:247]: "This allows the aggregation of fuel use to the utility level."

File: src/pudl/analysis/allocate_net_gen.py [hunk 11]
```diff
@@ -105,24 +105,37 @@
 """Data columns from generation_fuel_eia923 that are being allocated."""
 
 
-def allocate_gen_fuel_by_gen(pudl_out):
+# Two top-level functions (allocate & aggregate)
+
+
+def allocate_generation_fuel_by_generator_fuel_type(pudl_out, drop_interim_cols=True):
```
[the remark this answers is not in the document]
Comment (zaneselvans) [reply src/pudl/analysis/allocate_net_gen.py:111]: Having the `_by_` in the middle is helpful for distinguishing what is being allocated vs. how it is being allocated. It's missing from many of the other longer names over in the output modules.

File: src/pudl/analysis/allocate_net_gen.py [hunk 12]
```diff
@@ -235,25 +195,101 @@ def allocate_gen_fuel_by_gen_pm_fuel(gf, gen, gens, drop_interim_cols=True):
...
+
+    The generation_fuel_eia923 table includes net generation and fuel
+    consumption data at the plant/fuel type/prime mover level. The most
+    granular level of plants that PUDL typically uses is at the plant/generator
+    level. This function takes the plant/fuel type/prime mover level allocation,
+    aggregates it to the generator level and then denormalizes it to make it
+    more structurally in-line with the original generation_eia923 table.
+
+    Args:
+        pudl_out (pudl.output.pudltabl.PudlTabl): An object used to create the
```
Comment (zaneselvans) [inline src/pudl/analysis/allocate_net_gen.py:216]: You can put this type in the type hints up in the function call signature instead, so they're all in the same place.

**Review by zaneselvans — CHANGES_REQUESTED** [review]
I think that all the references to `fuel_type` in docstrings, comments, and names are probably out of date now, since those columns all refer to the set of codes defined by `energy_source_codes_eia` now, and we've renamed them to be called `energy_source_code` across the whole database (they were a heterogeneous mess before). If there is a fuel type column floating around, it's probably `fuel_type_code_pudl`.  This disconnect in naming could be confusing.

Having `DATA_COLS` as a module-level constant feels confusing -- it obscures what the core data is that's being manipulated in the module, and it's only used in a couple of places lower down.

Is there a reason why the columns that are enumerated in the various `IDX_` constants are ordered differently? For readability of the constants and also legibility of the dataframes they producing I think having a consistent ordering would be helpful.

There are so many different kinds of gen/gen fuel allocated tables floating around now that I think the functions / dataframes really need a strict naming convention and/or fuller docstrings indicating what the primary key & data columns are, and on what basis the data has been allocated. I got lost in the output functions, where there's old generic names, new specific names, the prime mover is often left out of the list of PK columns, different names for the accessor methods, the stored dataframes in the `self._dfs` dictionary, and the analysis functions which generate the dataframes.

File: src/pudl/output/pudltabl.py (hunk 5 above)
Comment (cmgosnell) [reply src/pudl/output/pudltabl.py:818]: is `gen_fuel_allocated_by_generator_eia923` better?

File: src/pudl/analysis/allocate_net_gen.py (hunk 8 above)
Comment (cmgosnell) [reply src/pudl/analysis/allocate_net_gen.py:212]: it means merge in all of the generator and unit and utility information. it is the same process that is used on the non-allocated/original generation table in the outputs

File: src/pudl/analysis/allocate_net_gen.py (hunk 12 above)
Comment (cmgosnell) [reply src/pudl/analysis/allocate_net_gen.py:216]: i'm getting circular references if i import PudlTabl into this module

**Review by zaneselvans — APPROVED** [review]

**Pull request merged**